Add Supermirror Polarizer Efficiency Algorithm

cailafinn opened the pull request:
Re-release as I can't push directly to @jclarkeSTFC's fork 
- #37180 

---

New algorithm to calculate the supermirror polarizer efficiency using a group workspace containing spin states, plus the efficiency of the analyser cell, calculated previously. This is part of the polarized SANS epic.

Output can either be to a workspace, a file, or both.

Fixes #36139

#### Summary of work
<!-- Please provide a short, high level description of the work that was done.
-->
- Added algorithm
- Added tests
- Added doc page

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

#### Further detail of work
<!-- Please provide a more detailed description of the work that has been undertaken.
-->
Calculates how the efficiency of a polarizer varies with wavelength. The ordering of the workspaces in ``InputWorkspace`` is taken from the ``SpinStates`` parameter, and the analyser efficiency, e_cell, is given by ``AnalyserEfficiency``.

The polarization of the polarizer, $P_{SM}$, is given by

$$
\huge P_{SM} = \frac{T_{00} - T_{01}}{2P_{cell}(T_{00} + T_{01})} = \frac{T_{00} - T_{01}}{2(2\epsilon_{cell} - 1)(T_{00} + T_{01})}
$$

Since the efficiency, $\epsilon_{SM}$ , is given by $\frac{1 + P_{SM}}{2}$, we have that

$$
\huge \epsilon_{SM} = \frac{1}{2} + \frac{T_{00} - T_{01}}{4(2\epsilon_{cell} - 1)(T_{00} + T_{01})}
$$

Error correction:

$$
\huge \sigma_{\epsilon_{SM}} = \sqrt{|\frac{\delta \epsilon_{SM}}{\delta T_{00}}|^2 * \sigma^2_{T_{00}} + |\frac{\delta \epsilon_{SM}}{\delta T_{01}}|^2 * \sigma^2_{T_{01}} + |\frac{\delta \epsilon_{SM}}{\delta \epsilon_{cell}}|^2 * \sigma^2_{\epsilon_{cell}}}
$$

where:

$$
\huge \frac{\delta \epsilon_{SM}}{\delta T_{00}} = \frac{T_{01}}{2(2\epsilon_{cell} - 1)(T_{00} + T_{01})^2}
$$

$$
\huge \frac{\delta \epsilon_{SM}}{\delta T_{01}} = \frac{-T_{00}}{2(2\epsilon_{cell} - 1)(T_{00} + T_{01})^2}
$$

$$
\huge \frac{\delta \epsilon_{SM}}{\delta \epsilon_{cell}} = \frac{T_{01} - T_{00}}{2(2\epsilon_{cell} - 1)^2(T_{00} + T_{01})}
$$



### To test:

<!-- Instructions for testing.
There should be sufficient instructions for someone unfamiliar with the application to test - unless a specific
reviewer is requested.
If instructions for replicating the fault are contained in the linked issue then it is OK to refer back to these.
-->
This will generate a simple example:
``` Python
wsPara = CreateSampleWorkspace('Histogram', Function='User Defined', UserDefinedFunction='name=UserFunction,Formula=0.5*exp(-0.0733*12*x*(1-0.9))',XUnit='Wavelength', xMin='1',XMax='8', BinWidth='1', NumBanks='1', BankPixelWidth='1')
wsPara1 = CloneWorkspace(wsPara)
wsAnti = CreateSampleWorkspace('Histogram', Function='User Defined', UserDefinedFunction='name=UserFunction,Formula=0.5*exp(-0.0733*12*x*(1+0.9))',XUnit='Wavelength', xMin='1',XMax='8', BinWidth='1', NumBanks='1', BankPixelWidth='1')
wsAnti1 = CloneWorkspace(wsAnti)

grp = GroupWorkspaces([wsPara,wsAnti,wsPara1,wsAnti1])
eCell = CreateSampleWorkspace('Histogram', Function='User Defined', UserDefinedFunction='name=UserFunction,Formula=(1 + tanh(0.0733 * 12 * x * 0.2))/2',XUnit='Wavelength', xMin='1',XMax='8', BinWidth='1', NumBanks='1', BankPixelWidth='1')

psm = PolarizerEfficiency(grp, eCell)
print("Polarizer efficiency at a wavelength of " + str(mtd['psm'].dataX(0)[3]) + " Å is " + str(mtd['psm'].dataY(0)[3]))
```

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

jclarkeSTFC committed:
Add algorithm, tests, and docs

Calculate polarizer efficiency using a group workspace with
4 runs for the 4 spin combinations, plus the efficiency of the
analyser cell, calculated previously.

jclarkeSTFC committed:
Fix doc test

jclarkeSTFC committed:
Update to use new formula for efficiency

jclarkeSTFC committed:
Pre-emptive attack on Coverity

jclarkeSTFC committed:
Tidy up code

Missing const in a few places, deleted unneeded methods and used the
operator overloads instead.

jclarkeSTFC committed:
Update formula for polarizer efficiency

Includes updating expected test results

jclarkeSTFC committed:
Fix cppcheck

jclarkeSTFC committed:
Update equations in docs

jclarkeSTFC committed:
Add option to save to file

You can now either have an output workspace, or save to a file, or both.

jclarkeSTFC committed:
Specify input is a workspace group

Co-authored-by: Caila Finn <47181718+cailafinn@users.noreply.github.com>

jclarkeSTFC committed:
Input may not be in the ADS

Co-authored-by: Caila Finn <47181718+cailafinn@users.noreply.github.com>

jclarkeSTFC committed:
Remove validator on input workspace

This validator does not work when the type is specified explicitly as
being a WorkspaceGroup, so instead we'll have to to check for wavelength
in validateInputs().

rbauststfc committed:
Add note about rebinning to documentation

rbauststfc committed:
Minor refactoring

cailafinn committed:
Remove calls to the ADS

RE #36139

cailafinn committed:
Remove requirement for four input workspaces in group

RE #36139

cailafinn committed:
Remove unecessary group validation and processing

If the property is defined such that it only allows workspace groups,
the group handling is performed implicitly.

RE #36139

cailafinn committed:
Check number of histograms in analyser & input match

RE #36139

cailafinn committed:
Ensure analyser workspace is a MatrixWorkspace

RE #36139

cailafinn committed:
Use correct error calculation

RE #36139

cailafinn committed:
Improve error calculation performance

Workspace arithmetic calculations get expensive when there are a lot of
bins. Performing the calculations directly on the Y and E arrays is
much more performant.

RE #36139

cailafinn committed:
Add error calculation te equations to the docs

RE #36139

cailafinn committed:
Fix doctests

We're limiting to one spectra per workspace now, so restrict the example
workspaces in the test.

RE #36139

cailafinn committed:
Remove unneeded power method

It just called the power algorithm, but we're not using workspace
arithmetic anymore.

RE #36139

rbauststfc commented on Framework/Algorithms/src/PolarizationCorrections/PolarizerEfficiency.cpp:
```diff
@@ -0,0 +1,201 @@
+  const WorkspaceGroup_sptr &groupWorkspace = getProperty(PropertyNames::INPUT_WORKSPACE);
+  const auto spinConfigurationInput = getPropertyValue(PropertyNames::SPIN_STATES);
+
+  const auto &t01Ws = PolarizationCorrectionsHelpers::workspaceForSpinState(groupWorkspace, spinConfigurationInput,
```
I think we need some validation to check that `01` and `00` are included in the spin state configuration input, even if there are more than two configurations supplied. I tested using an input workspace group of 4 workspaces and with spin configuration input `11, 00` and it ran through without error. The calculated efficiencies were different compared to when I set the spin configuration input to `01, 00`. I think we possibly want to validate that whatever number of spin configurations has been passed in matches the number of workspaces in the group. But at the very least, I think we need to check that we have the required `01` and `00` for the calculation.

In my first invalid test I was surprised not to get an error from `PolarizationCorrectionsHelpers::workspaceForSpinState`, because it's looking for spin state `01` that wasn't passed in, so that could do with a look.

cailafinn replied:
Turns out the source of this is due to the way the Helper function works. If it fails to find a match for a given spin state (01, say) it returns the index 1 beyond the last index in the list of spin states. This means that if the sizes do not match, the helper can return an invalid index for the spin states that _is_ valid for the workspace group (provided it is larger). I'll make a fix for both issues and validate for both, as it's just confusing otherwise.

rbauststfc reviewed (commented):
Thanks very much for picking this one up. There are a few bits that I think could do with a look, but thanks very much for spending the time investigating the approach to the error calculation to optimize the performance there. Some very useful insights from that!

cailafinn commented on Framework/Algorithms/src/PolarizationCorrections/PolarizerEfficiency.cpp:
```diff
@@ -0,0 +1,201 @@
+  saveAlg->execute();
+}
+
+MatrixWorkspace_sptr PolarizerEfficiency::convertToHistIfNecessary(const MatrixWorkspace_sptr ws) {
```
@jclarkeSTFC Do you know why this is? I don't know if you came across something during testing?

jclarkeSTFC replied:
For the shared pointers, passing by reference can apparently mess with how the shared pointer keeps track of how many times it is referenced, e.g. see [here](https://stackoverflow.com/questions/8385457/should-i-pass-a-shared-ptr-by-reference). Plus if we pass by reference we still have to copy a pointer so it's not much of an efficiency saving.

If they're always going to pass in histogram workspaces then no need for the conversion, but then you probably want some validation on the input.

cailafinn committed:
Refactor workspace validation

RE #36139

cailafinn committed:
Fix eCell delta calculation

RE #36139

cailafinn committed:
Add references to documentation

RE #36139

cailafinn committed:
Validate for matching input and spinstate length

RE #36139

cailafinn committed:
Fix incorrect signing

Also adds some documentation to the function.

RE #36139

rbauststfc commented on Framework/Algorithms/src/PolarizationCorrections/PolarizerEfficiency.cpp:
```diff
@@ -0,0 +1,204 @@
+  const auto &t00Ws = PolarizationCorrectionsHelpers::workspaceForSpinState(inputWorkspace, spinConfigurationInput,
+                                                                            SpinStateValidator::ZERO_ZERO);
+  if (t01Ws == nullptr || t00Ws == nullptr) {
+    errorList[PropertyNames::SPIN_STATES] = "The required spin configurations (00, 01) could not be found in the given"
```
I unexpectedly get this as a second validation error message if I pass in spin configuration `00,01,10,11` with a group of table workspaces (the crash we were getting with this type of input workspace group is now resolved, though, so that's great). I think this might be resolved if we were able to use `PolarizationCorrectionsHelpers::indexOfWorkspaceForSpinState` directly in the validation here rather than `PolarizationCorrectionsHelpers::workspaceForSpinState`?

rbauststfc reviewed (commented):
Validation is looking much more robust now and the implementation of the error calculation all looks good, so thanks very much for all of that. I've flagged a couple of bits I noticed and I'm afraid there's also an update on requirements regarding the rebinning of the input efficiency workspace that will need a slight change, so apologies about that.

cailafinn commented on Framework/Algorithms/src/PolarizationCorrections/PolarizerEfficiency.cpp:
```diff
@@ -0,0 +1,204 @@
+    errorList[propertyName] = "All input workspaces must be of type MatrixWorkspace.";
+    return;
+  }
+  if (ws->getNumberHistograms() != 1) {
```
There is a benefit to adding as many as possible to the error list before returning,  as the red asterisks in the Algorithm Dialog are set based on this. So it makes it more clear there that there are multiple issues.

rbauststfc replied:
I think maybe my comment is a bit confusing, sorry. At this point in the code we're in helper method `validateInputWorkspace` to check for errors with a single property, rather than in the `validateInputs` method. If we return early from `validateInputWorkspace` then I think we would still check the rest of the properties. I'm just thinking that when we're validating a single property (so we're only going to get one asterisk out of it, regardless of the number of problems found) we can technically stop checking as soon as the first problem is encountered.

cailafinn committed:
Optionally return an index for a given spin state

We're not certain an index will be found, so make use of std::optional
to determine if we found one or not.

RE #36139

cailafinn committed:
Amend release note locations

RE #36139

cailafinn committed:
Remove silent rebinning

RE #36139

cailafinn committed:
Add test for invalid analyser ws

RE #36139

cailafinn committed:
Bail out of validation early if non-matrix workspaces in group

RE #36139

rbauststfc committed:
Update documentation

Remove obsolete reference to rebinning step in docs and fix example that
no longer works since automatic rebinning has been removed.

RE mantidproject#36139

rbauststfc committed:
Add PolarizationCorrectionHelpers test

Add new test for when no workspace is found for the requested spin
state.

RE mantidproject#36139

rbauststfc committed:
Fix crash and minor validation refactor

Prevent crash when input analyser workspace is not of the correct type.
Also refactor to use indexOfWorkspaceForSpinState directly in
validateInputs.

RE mantidproject#36139

rbauststfc reviewed (dismissed):
I've done a thorough test of the functionality, including checking the validation, and that's all looking good. Code changes are also looking good and I'm happy that all previous review comments have been addressed.

I think we could do with some more tests for the `PolarizationCorrectionHelpers` methods, as only one of the methods is currently being tested. However I have opened a separate maintenance issue to look at that (#37749), as it's beyond the scope of this PR.

On re-review I did notice a few small issues, but I'm keen to get this one into a Nightly in time for Thursday if possible, so I decided to address those myself as Caila is currently on leave. The last three commits contain these changes, with an explanation in the commit message of what is being addressed. I've discussed with Sarah and because these changes are fairly contained, it seemed unnecessary to get a new first reviewer. I'm therefore approving and asking whoever gatekeepers this to pay particular attention to the three commits that I've added.

RichardWaiteSTFC commented on Framework/Algorithms/src/PolarizationCorrections/PolarizerEfficiency.cpp:
```diff
@@ -0,0 +1,201 @@
+    const auto &delta00 = (t01Y[i]) / (2 * (2 * effCellY[i] - 1) * pow(t00Y[i] + t01Y[i], 2));
+    const auto &delta01 = (-1 * t00Y[i]) / (2 * (2 * effCellY[i] - 1) * pow(t00Y[i] + t01Y[i], 2));
+    const auto &deltaEffCell = (t01Y[i] - t00Y[i]) / (pow(2 * (2 * effCellY[i] - 1), 2) * (t00Y[i] + t01Y[i]));
+    effPolarizerE[i] = sqrt(pow(abs(delta00), 2) * pow(t00E[i], 2) + pow(abs(delta01), 2) * pow(t01E[i], 2) +
```
Yes no need for `abs` as you say. Also I thought `pow(num, 2)` was slower than `num * num` - but with equations this long that might make them difficult to read! I think you could at least call `pow` fewer times e.g. 
```suggestion
    effPolarizerE[i] = sqrt(pow(delta00*t00E[i], 2) + pow(delta01*t01E[i], 2) +
```
It does look like some of the terms are computed in multiple places (e.g. `(2 * effCellY[i] - 1)` and `t00Y[i] + t01Y[i]`) - I don't know whether it's worth putting them in variables to make this obvious (I don't know whether that actually improves speed, I guess not as there's some overhead to assigning the variable). It might be more readable?

rbauststfc replied:
Thanks for this Richard, some useful points that I'll take a look at. I hadn't realised (remembered? - maybe at some point I would have come across this a long time ago 🙂) that mathematically $x^2 * y^2$ is equivalent to $(x*y)^2$, so that's really useful to know.

RichardWaiteSTFC reviewed (commented):
Looks good - I had some comments but If you need this merged they can be addressed later if you like (or not at all!)? Just let me know!

rbauststfc committed:
Minor refactor of error calculation

Changes to the error calculation to cut down on duplicate calculations,
reduce use of pow(num,2) and improve readability.

RE mantidproject#36139

rbauststfc commented:
I've made the suggested changes to the error calculation. Only a very small speed improvement, but I think it aids readability. I'll approve this again when tests have passed and then it will be ready for you to gatekeeper again @RichardWaiteSTFC. Thanks very much for your help with this one.

rbauststfc reviewed (approved):
Approving following the amendment to the error calculation. I'm conscious I'm approving my own change here, but I'm sure @RichardWaiteSTFC will review the changes carefully as part of gatekeepering.
